dot/sync/disjoint_block_set_test.go

```diff
@@ -0,0 +1,331 @@
+		block *types.Block
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		err    error
+	}{
+		{
+			name:   "add block beyond capacity",
+			fields: fields{},
+			args: args{block: &types.Block{
+				Header: types.Header{
+					Number: 1,
+				},
+			}},
+			err: errors.New("cannot add block; set is at capacity"),
+		},
+		{
+			name: "add block",
+			fields: fields{
+				limit:            1,
+				blocks:           make(map[common.Hash]*pendingBlock),
+				timeNow:          time.Now,
+				parentToChildren: make(map[common.Hash]map[common.Hash]struct{}),
+			},
+			args: args{block: &types.Block{
+				Header: types.Header{
+					Number: 1,
+				},
+			}},
+			err: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := &disjointBlockSet{
+				limit:            tt.fields.limit,
+				blocks:           tt.fields.blocks,
+				parentToChildren: tt.fields.parentToChildren,
+				timeNow:          tt.fields.timeNow,
+			}
+			err := s.addBlock(tt.args.block)
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_disjointBlockSet_addHeader(t *testing.T) {
+	t.Parallel()
+
+	type fields struct {
+		limit            int
+		blocks           map[common.Hash]*pendingBlock
+		parentToChildren map[common.Hash]map[common.Hash]struct{}
+		timeNow          func() time.Time
+	}
+	type args struct {
+		header *types.Header
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		err    error
+	}{
+		{
+			name:   "add header beyond capactiy",
+			fields: fields{},
+			args: args{header: &types.Header{
+				Number: 1,
+			}},
+			err: errors.New("cannot add block; set is at capacity"),
+		},
+		{
+			name: "add header",
+			fields: fields{
+				blocks:           make(map[common.Hash]*pendingBlock),
+				limit:            1,
+				timeNow:          time.Now,
+				parentToChildren: make(map[common.Hash]map[common.Hash]struct{}),
+			},
+			args: args{header: &types.Header{
+				Number: 1,
+			}},
+			err: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := &disjointBlockSet{
+				limit:            tt.fields.limit,
+				blocks:           tt.fields.blocks,
+				parentToChildren: tt.fields.parentToChildren,
+				timeNow:          tt.fields.timeNow,
+			}
+			err := s.addHeader(tt.args.header)
+			if tt.err != nil {
+				assert.EqualError(t, err, tt.err.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_disjointBlockSet_clearBlocks(t *testing.T) {
+	t.Parallel()
+
+	testBlock := &pendingBlock{
+		clearAt: time.Now(),
+	}
+	type fields struct {
+		limit            int
+		blocks           map[common.Hash]*pendingBlock
+		parentToChildren map[common.Hash]map[common.Hash]struct{}
+		timeNow          func() time.Time
+	}
+	tests := []struct {
+		name      string
+		fields    fields
+		remaining int
+	}{
+		{
+			name: "base case",
```

Comment:
can you add a case where there are some remaining blocks left over?

Reply:
Yes, added test case.